### ISSUE TITLE
[idea] On-demand loading of JavaScript files (or prevent them from loading at all).

### ISSUE THREAD

**Issue by trusktr** [meteor/meteor#5883]
It'd be really great, now that we have more control of source due to the new `modules` system and our new ability to make entry points (`main.js` files), if we could specify which files (entry points or regular non-module files) for Meteor _not_ to load, so that we can manually load them later. Meteor would simply need to serve a URL for the specified file.

I posted more details about how the configuration, etc, might look like in [this forum post](https://forums.meteor.com/t/idea-manual-control-of-loading-of-entrypoints-in-meteor-1-3/15368).

**Comment by benjamn** [comment]
This is a great idea, and something we definitely hope to investigate after 1.3 (for like 1.3.1 or so).

**Comment by trusktr** [comment]
Sweet! Would Meteor get it's own API for loading files? Maybe something like `Meteor.load('/app/path/to/file.js')`? I wonder what the fate of `System` (i.e. what SystemJS currently "shims") will be? If `System` will for sure to become an official part of browser APIs at some point in the future, would it make sense to use SystemJS now?

**Comment by trusktr** [comment]
@benjamn Could `meteorInstall` register modules in such a way that shared (built, minified) modules will be fetched in parallel only the first time they are required (cached after that)? That would make the module system even more killer.), for things like conditional loading based routes for example.

**Comment by mitar** [comment]
See my solution here: #6203

**Comment by mitar** [comment]
(@trusktr, it feels like we are solving similar issues.)

**Comment by trusktr** [comment]
@mitar That's a nice concept, but I think it is a temporary hack, as it won't fetch scripts all at once, waiting for the sync load of each script between network requests. I think an async solution would be best for performance. I imagine if we just prevent Meteor from writing the `<script>` tag based on a user config match ([for example](https://forums.meteor.com/t/idea-manual-control-of-loading-of-entrypoints-in-meteor-1-3/15368/2?u=joe)) then we provide a simple method to load by URL:

```js
Meteor.load(URL).then(...) // no call-back returns promise.
// or
Meteor.load(URL, callback) // call-back, no promise
```

That method will need to rely on @benjamn's [install](https://www.npmjs.com/package/install) (`meteorInstall`?) so that the module can be previously registered with its dependencies but loaded (along with dependencies) only when called with `Meter.load(url)` instead of included inline.

This is just theory though. I would love to work on this, but I have no time right now, as now I need to go work on what I'm being paid for. Wish I could!

**Comment by mitar** [comment]
> That's a nice concept, but I think it is a temporary hack, as it won't fetch scripts all at once, waiting for the sync load of each script between network requests.

Oh, that is just a question how you want them to be loaded. Instead of resolving them recursively, you could just load all of them at once, and then `Promise.all` on them.

In my real code I use Meteor manifest file to get information about all dependencies in advance and can then request all dependencies as needed.

**Comment by mitar** [comment]
> In my real code I use Meteor manifest file to get information about all dependencies in advance and can then request all dependencies as needed.

So `dependenciesReady` helps you start loading scripts before you have a manifest file loaded, and once you have a manifest file loaded you can start loading things synchronously.

**Comment by mitar** [comment]
I think this has been implemented in Meteor 1.5?
